cpp/ql/src/Likely Bugs/Likely Typos/UsingStrcpyAsBoolean.ql

```diff
@@ -1,77 +1,91 @@
-  and msg = "Return Value of " + func.getTarget().getQualifiedName() + " used as boolean."
+predicate isStringCopyCastedAsBoolean(FunctionCall func, Expr expr1, string msg) {
+  DataFlow::localFlow(DataFlow::exprNode(func), DataFlow::exprNode(expr1)) and
+  isBoolean(expr1.getConversion*()) and
+  isStringComparisonFunction(func.getTarget().getName()) and
+  msg = "Return value of " + func.getTarget().getName() + " used as Boolean."
 }
 
-predicate isStringCopyUsedInLogicalOperationOrCondition( FunctionCall func, Expr expr1, string msg ) {
-    isStringComparisonFunction( func.getTarget().getQualifiedName() )
-    and ((( 
-      // it is being used in an equality or logical operation 
-      exists( EqualityOperation eop |
-        eop = expr1
-        and func = eop.getAChild()
+predicate isStringCopyUsedInLogicalOperationOrCondition(FunctionCall func, Expr expr1, string msg) {
+  isStringComparisonFunction(func.getTarget().getName()) and
+  (
+    (
+      // it is being used in an equality or logical operation
+      exists(EqualityOperation eop |
+        eop = expr1 and
+        func = eop.getAnOperand()
       )
-      or exists( UnaryLogicalOperation ule |
-        expr1 = ule
-        and func = ule.getAChild()
+      or
+      exists(UnaryLogicalOperation ule |
+        expr1 = ule and
+        func = ule.getOperand()
       )
-      or exists( BinaryLogicalOperation ble |
-        expr1 = ble
-        and func = ble.getAChild()
+      or
+      exists(BinaryLogicalOperation ble |
+        expr1 = ble and
+        func = ble.getAnOperand()
       )
-    )
-    and msg = "Return Value of " + func.getTarget().getQualifiedName() + " used in a logical operation."
-    )
+    ) and
+    msg = "Return value of " + func.getTarget().getName() +
+        " used in a logical operation."
     or
-      exists( ConditionalStmt condstmt |
-      condstmt.getAChild() = expr1 |
-      // or the string copy function is used directly as the conditional expression
-      func = condstmt.getChild(0)
-      and msg = "Return Value of " + func.getTarget().getQualifiedName() + " used directly in a conditional expression."
-    ))
+    // or the string copy function is used directly as the conditional expression
+    (
+      exists(ConditionalStmt condstmt |
+        func = condstmt.getControllingExpr() and
+        expr1 = func
+      )
+      or
+      exists(ConditionalExpr ce |
+        expr1 = ce and
+        func = ce.getCondition()
+      )
+    ) and
+    msg = "Return value of " + func.getTarget().getName() +
+        " used directly in a conditional expression."
+  )
 }
 
 from FunctionCall func, Expr expr1, string msg
-where 
-  ( isStringCopyCastedAsBoolean(func, expr1, msg) and 
-    not isStringCopyUsedInLogicalOperationOrCondition(func, expr1, _)
+where
+  (
+    isStringCopyCastedAsBoolean(func, expr1, msg) and
+    not isStringCopyUsedInLogicalOperationOrCondition(func, _, _)
```

Comment:
Despite this line it's still possible to get duplicate results, e.g.
```
bool r;
r = (bool)strcpy(buf, "world"); // two results on this line
if (r) // one result on this line
{
}
```

Reply:
Did you test that? I don't see how that's possible. I'd expect one result only from your test case: on the `if (r)` line.

Reply:
I did test that.  I think there were multiple results from `isStringCopyCastedAsBoolean` (rather than an overlap between the cases) because there are multiple places where the expression has a boolean value.

Reply:
It makes sense that `isStringCopyCastedAsBoolean` can have multiple results for the same `strcpy` call. There's not necessarily a single correct location to report in the general case.

Reply:
Fair enough.  I was wondering about restricting it to the point where a non-boolean is cast to boolean - though it will still be possible to cast it back and forth.